File: RmvDup.js
{/*https://leetcode.com/problems/contains-duplicate/

Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.

Example 1:

Input: nums = [1,2,3,1]
Output: true
Explanation:
The element 1 occurs at the indices 0 and 3.
*/}

//BRUTE-FORCE APPROACH
function removeDuplicates(arr) {
    let set = new Set(arr);
    let uniqueArr = Array.from(set);
    for (let i = 0; i < uniqueArr.length; i++) {
      arr[i] = uniqueArr[i];
    }
    return uniqueArr.length;
}
//Time Complexcity:- O(NlogN)+O(N)


//OPTIMAL APPROACH
function removeDuplicates(arr) {
    let i = 0;
    for (let j = 1; j < arr.length; j++) {
      if (arr[i] !== arr[j]) {
        i++;
        arr[i] = arr[j];
      }
    }
    return i + 1;
  }

//Time Complexcity:- O(N)

File: RmvDup.py
# ##https://leetcode.com/problems/contains-duplicate/

# Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.

# Example 1:

# Input: nums = [1,2,3,1]
# Output: true
# Explanation:
# The element 1 occurs at the indices 0 and 3.


array = [1,2,46,32,98,61,34,46]

#BRUTE-FORCE APPROACH
def brute_force_duplicate_search(array):
    for i in range(len(array)-1):
        for j in range(i+1,len(array)):
            if array[i] == array[j]:
                return True
    return False

print(brute_force_duplicate_search(array))
#Time Complexcity:- O(NlogN)+O(N)

#OPTIMAL APPROACH
def smart_duplicate_search(array):
    dictionary = dict()
    if len(array)<2:
        return False
    else:
        for i in range(len(array)):
            if array[i] in dictionary:
                return True
            else:
                dictionary[array[i]] = True
    return False

print(smart_duplicate_search(array))